browser/greaselion/greaselion_tab_helper.cc

```diff
@@ -0,0 +1,89 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/greaselion/greaselion_tab_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "base/bind_helpers.h"
+#include "base/strings/utf_string_conversions.h"
+#include "brave/browser/greaselion/greaselion_service_factory.h"
+#include "brave/common/brave_isolated_worlds.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service_factory.h"
+#include "brave/components/greaselion/browser/greaselion_service.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+using greaselion::GreaselionService;
+using greaselion::GreaselionServiceFactory;
+
+namespace greaselion {
+
+GreaselionTabHelper::GreaselionTabHelper(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents->GetBrowserContext());
+  if (profile) {
+    brave_rewards::RewardsService* rewards_service =
+        brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+    if (rewards_service)
+      rewards_service->AddObserver(this);
+  }
+}
+
+GreaselionTabHelper::~GreaselionTabHelper() {
+  if (!web_contents())
+    return;
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  if (!profile)
+    return;
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (rewards_service)
+    rewards_service->RemoveObserver(this);
+}
+
+void GreaselionTabHelper::DocumentLoadedInFrame(
+    content::RenderFrameHost* render_frame_host) {
+  const GURL& url = render_frame_host->GetLastCommittedURL();
+  std::vector<std::string> scripts;
+  GreaselionService* greaselion_service =
+      GreaselionServiceFactory::GetForBrowserContext(
+          web_contents()->GetBrowserContext());
+  if (!greaselion_service)
+    return;
+  if (!greaselion_service->ScriptsFor(url, &scripts))
+    return;
+
+  for (auto script : scripts) {
+    render_frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(script), base::DoNothing(),
+        ISOLATED_WORLD_ID_GREASELION);
+  }
+}
+
+void GreaselionTabHelper::OnRewardsMainEnabled(
```

Comment:
Oh, actually my earlier comment missed something here. This still puts control of feature enable/disable in greaselion instead of rewards service. I think we want a RewardsServiceObserver in the rewards component for this so it maintains control.

Reply:
I would suggest GreaselionRewardsServiceObserver owned by RewardsServiceImpl (similar to extension_rewards_service_observer_). That also takes care of the https://github.com/brave/brave-core/pull/2312/files#r292231928